site/en/docs/extensions/reference/storage/index.md

```diff
@@ -81,50 +82,115 @@ tubes can be filled, and if they are filled when you put your message in, it get
-  chrome.storage.sync.set({'value': theValue}, function() {
-    // Notify that we saved.
-    message('Settings saved');
-  });
-}
+});
 ```
 
-If you're interested in tracking changes made to a data object, you can add a listener to its
-`onChanged` event. Whenever anything changes in storage, that event fires. Here's sample code to
-listen for saved changes:
+We can take this idea even further. In this example we have an [options page][options-page] that
+allows the user to toggle a "debug mode" in the extension (implementation not shown here). Changes
+to this setting are immediately saved to sync storage by the options page and the background script uses `storage.onChanged` to apply the setting as soon as possible.
 
 ```js
-chrome.storage.onChanged.addListener(function(changes, namespace) {
-  for (var key in changes) {
-    var storageChange = changes[key];
-    console.log('Storage key "%s" in namespace "%s" changed. ' +
-                'Old value was "%s", new value is "%s".',
-                key,
-                namespace,
-                storageChange.oldValue,
-                storageChange.newValue);
+//// options.html ////
+
+<script defer src="options.js"></script>
+<form id="optionsForm">
+  <label for="debug">
+    <input type="checkbox" name="debug" id="debug">
+    Enable debug mode
+  </label>
+</form>
+
+//// options.js ////
+
+// In-page cache of the user's options
+let options = {};
+
+// Initialize the form with the user's option settings
+chrome.storage.sync.get('options', (data) => {
+  Object.assign(options, data?.options);
+  optionsForm.debug.checked = data?.options?.debug;
+});
+
+// Immediately persist options changes
+optionsForm.debug.addEventListener('change', (event) => {
+  options.debug = event.target.checked;
+  chrome.storage.sync.set({options});
+});
+
+//// background.js ////
+
+// Watch for changes to the user's options & apply them
+chrome.storage.onChanged.addListener((changes, area) => {
+  if (area === 'sync' && changes.options) {
+    console.log('enable debug mode?', !!changes.options?.newValue?.debug);
+    setDebugMode(!!changes.options?.newValue?.debug);
   }
 });
 ```
 
-[1]: https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage
-[2]: /docs/extensions/mv2/manifest/incognito
-[3]: /docs/extensions/mv2/manifest/storage
-[4]: /docs/extensions/mv2/tabs
-[5]: #property-sync
-[6]: #property-local
+Since service workers are not always running, Manifest V3 extensions sometimes need to
+asynchronously load data from storage before they execute their event handlers. To do this, the
+below snippet uses an async `action.onClicked` event handler that waits for the `storageCache`
+global to be populated before executing it's logic.
+
+```js
+//// background.js ////
+
+// Where we will expose all the data we retrieve from storage.sync
+let storageCache = {};
+// Asynchronously retrieve data from storage.sync, then cache it
+let initStorageCache = getAllStorageSyncData().then(items => {
```

Comment:
While the idea looks promising on paper but it doesn't actually work as the API request is made in the same initial task shared by the onClicked listener so this code doesn't make any difference to just reading the storage inside onClicked.

Reply:
I guess this idea can help when performing a heavy synchronous operation at start before `chrome` listener runs:

```js
preloadStorage();
superHeavySynchronousOperation(); // or importScripts('super-big-library.js');
chrome.action.onClicked.addListener(fnUsingPreloadedStorage);
```

An example of such operation would be loading of dependencies but unfortunately this won't help with modern complex apps produced by webpack and such because the main part of a script runs only after all its complex/big dependencies were executed, meaning that storage preloader would run at the end of the task when it's already useless...

In such cases the solution would be to write a wrapper bg.js manually:

```js
try {
  importScripts('preload-storage.js', 'bg-main-webpack-chunk.js');
} catch (e) {
  console.error(e);
}
```

...where preload-storage.js either uses a global `const` or a global `self.something`.

I suggest describing this in detail, maybe even make a demo webpack-based extension, otherwise the idea won't be used properly.

Reply:
I don't really understand your second comment here @tophf. Service Workers currently do not run in module mode, so there's no way to do top-level work (no top-level `await`). No amount of rearranging dependencies can change that.

As for your first comment: yes, the loader will happen at the same time. But in general this approach is fine because it's not clear if there's going to be additional listeners added.

I'm generally not a huge fan of making event listeners async, as it gives folks a false sense of—they can do async things without worrying about exceptions—but I'm growing a bit more fine with it.

Reply:
> I don't really understand your second comment here @tophf. Service Workers currently do not run in module mode, so there's no way to do top-level work (no top-level await). 

They will soon but my second comment is not related to that.

> No amount of rearranging dependencies can change that.

It's about the multi-process architecture of the API: the earlier the request is sent to the browser process the sooner it will be processed **there**. So, after the API request is sent, the extension process will be busy compiling and running the code of its synchronously executed dependencies/libraries. For example, it may take 100ms or even 1000ms if there's a lot of code so by the time the listener runs, the storage will be ready in the browser process and quickly transferred to the extension process.

Reply:
Thanks for the explanation, and I see your point. However, for a simple example page—this is getting a bit in-depth.

I'd prefer module service workers (I suspect they'll work fine—they won't complete loading until the async work is done, so nothing will try to answer the incoming event). It seems like Chrome is about to make them stable. The question though is it worth updating to that model ASAP or showing the older approach for longer.

Reply:
> While the idea looks promising on paper but it doesn't actually work as the API request is made in the same initial task shared by the onClicked listener so this code doesn't make any difference to just reading the storage inside onClicked.

I believe you are mistaken.

First, there is no guarantee that the storage get call will complete before an event handler is executed. In many cases that may be true for the reasons you described, but it is not always true. In my local testing I observed a number of cases where `storageCache` was empty at the beginning of the `action.onClicked` event handler call but it was populated after awaiting the resolution of `initStorageCache`.

To reproduce this, use the below code and allow ~30 second after the service worker is inactive before clicking the browser action button. Then, open DevTools for the service worker and observe the logs.

```js
let storageCache = {};
let initStorageCache = (() => {
  return new Promise((resolve) => {
    chrome.storage.sync.get(null, (data) => {
      Object.assign(storageCache, data);
      resolve(storageCache);
    });
  });
})();

chrome.runtime.onInstalled.addListener(() => {
  chrome.storage.sync.set({foo: 'bar'});
});

chrome.action.onClicked.addListener(async (tab) => {
  console.log(initStorageCache, storageCache.foo);
  await initStorageCache;
  console.log(initStorageCache, storageCache.foo);
});
```

Second, even if the `initStorageCache` promise is resolved before the event handler is executed, that doesn't mean "doesn't actually work", just that in that case it is superfluous. The pattern I'm demonstrating here guarantees consistent behavior. We do not know where these samples will end up being used, so we should be resilient against slow disks, large amounts of stored data, etc.

> I'm generally not a huge fan of making event listeners async, as it gives folks a false sense of—they can do async things without worrying about exceptions—but I'm growing a bit more fine with it.

I'm for it if only because there's no other way to await promise resolution. I'm bullish on the pattern of deferring the execution of an event listener's body until you're sure all work necessary to handle the event has completed.  To your point, though, it may be best to include a try/catch to better set expectations.

```js
chrome.action.onClicked.addListener(async (tab) => {
  try {
    await initStorageCache;
  } catch (e) {
    // handle error that occurred during storage initialization
  }
  // normal action handler logic
});
```

I'll integrate the above block.

> However, for a simple example page—this is getting a bit in-depth.

IMO the discussion is all relevant. I'm not just aiming for these snippets to be useful in isolation, I also want them to demonstrate good development practices and be usable in a variety of existing extensions. Even if we can't detail ever decision, we can still demonstrate resilient patterns.

> The question though is it worth updating to that model ASAP or showing the older approach for longer.

Generally speaking, my approach has been to target the capabilities of Chrome stable and call out where they're not backwards compatible & why. Based on the timing of the CL that landed SW module support, I don't expect extensions will support module SWs until M91 at the earliest.

Reply:
I'm not mistaken but it seems like I didn't explain my point well. I didn't say that the storage **call** will complete, but the internal implementation of the storage which runs inside the separate process (the browser process).

To clarify, there are two observations I made. The first one is that the idea of preloading the storage at the beginning of the script has exactly the same performance as reading the storage inside the listener so it's useless. And then I amended this observation by finding a case where such preloading will be useful.

Reply:
> Oh and another benefit of preloading would be the case where multiple API events are registered so each listener can start with await initStorageCache

Yep, that's exactly the use case I had in mind. 

This is actually a simplification of a storage library I've been toying with to provide developers with a write-through cache backed by IndexedDB or the Storage API. The goal is to make it so devs can read/write data on a `storage` object they import and it handles the Storage API calls behind the scenes. I'm aiming for the ergonomics of a simple global object with the resilience of the Storage API.